gobblin-core/src/main/java/gobblin/writer/SimpleDataWriter.java

```diff
@@ -0,0 +1,170 @@
+      toWrite[toWrite.length - 1] = recordDelimiter.get();
+    }
+    if (prependSize) {
+      long recordSize = toWrite.length;
+      ByteBuffer buf = ByteBuffer.allocate(Long.SIZE / Byte.SIZE);
+      buf.putLong(recordSize);
+      toWrite = ArrayUtils.addAll(buf.array(), toWrite);
+    }
+    this.stagingFileOutputStream.write(toWrite);
+    bytesWritten += (toWrite.length);
+    recordsWritten++;
+  }
+
+  /**
+   * Commit the data written to the final output file.
+   *
+   * @throws java.io.IOException if there is anything wrong committing the output
+   */
+  @Override
+  public void commit() throws IOException {
+    this.close();
+    if (!this.fs.exists(this.stagingFile)) {
+      throw new IOException(String.format("File %s does not exist", this.stagingFile));
+    }
+
+    LOG.info(String.format("Moving data from %s to %s", this.stagingFile, this.outputFile));
+    // For the same reason as deleting the staging file if it already exists, deleting
+    // the output file if it already exists prevents task retry from being blocked.
+    if (this.fs.exists(this.outputFile)) {
+      LOG.warn(String.format("Task output file %s already exists", this.outputFile));
+      HadoopUtils.deletePath(this.fs, this.outputFile, false);
+    }
+    HadoopUtils.renamePath(this.fs, this.stagingFile, this.outputFile);
+  }
+
+  /**
+   * Cleanup context/resources.
+   *
+   * @throws java.io.IOException if there is anything wrong doing cleanup.
+   */
+  @Override
+  public void cleanup() throws IOException {
+    if (this.fs.exists(this.stagingFile)) {
+      HadoopUtils.deletePath(this.fs, this.stagingFile, false);
+    }
+  }
+
+  /**
+   * Get the number of records written.
+   *
+   * @return number of records written
+   */
+  @Override
+  public long recordsWritten() {
+    return this.recordsWritten;
+  }
+
+  /**
+   * Get the number of bytes written.
+   * <p>
+   * This method should ONLY be called after {@link gobblin.writer.DataWriter#commit()}
+   * is called.
+   * </p>
+   *
+   * @return number of bytes written
+   */
+  @Override
+  public long bytesWritten() throws IOException {
+    if (this.closed) {
+      return this.bytesWritten;
+    } else {
+      return 0;
+    }
+  }
+
+  /**
+   * Closes this stream and releases any system resources associated
+   * with it. If the stream is already closed then invoking this
+   * method has no effect.
+   * <p/>
+   * <p> As noted in {@link AutoCloseable#close()}, cases where the
+   * close may fail require careful attention. It is strongly advised
+   * to relinquish the underlying resources and to internally
+   * <em>mark</em> the {@code Closeable} as closed, prior to throwing
+   * the {@code IOException}.
+   *
+   * @throws java.io.IOException if an I/O error occurs
+   */
+  @Override
+  public void close() throws IOException {
+    if (stagingFileOutputStream != null && !this.closed) {
+      stagingFileOutputStream.flush();
```

Comment:
If `flush` throws an `IOException`, then `close` will never be called. Better to do:
```
try {
  this.stagingFileOutputStream.flush();
} finally {
  this.stagingFileOutputStream.close();
}
```